pkg/utils/retry/retry.go

```diff
@@ -65,6 +66,12 @@ func (s *Strategy[R]) Do(ctx context.Context, lggr logger.Logger, fn func(ctx co
 			return result, nil
 		}
 
+		// Handle permanent errors without retrying.
+		var permanent *PermanentError
+		if errors.As(err, &permanent) {
+			return result, err
```

Comment:
Oh should it be unwrapped?
```suggestion
		var permanent *PermanentError
		if errors.As(err, &permanent) {
			return result, permanent.Unwrap()
```